lib/TemplatedPathPlugin.js

```diff
@@ -75,130 +90,187 @@ const getReplacer = (value, allowEmpty) => {
+				// @ts-ignore
+				chunk.contentHashWithLength[contentHashType]) ||
+			undefined;
+
+		if (typeof path === "function") {
+			path = path(data);
+		}
+
+		if (
+			data.noChunkHash &&
+			(REGEXP_CHUNKHASH_FOR_TEST.test(path) ||
+				REGEXP_CONTENTHASH_FOR_TEST.test(path))
+		) {
+			throw new Error(
+				`Cannot use [chunkhash] or [contenthash] for chunk in '${path}' (use [hash] instead)`
+			);
+		}
+
+		return (
+			path
+				.replace(REGEXP_ID, replacer(id))
+				.replace(REGEXP_NAME, replacer(name))
+				.replace(
+					REGEXP_HASH,
+					hashLength(replacer(data.hash), data.hashWithLength)
+				)
+				.replace(
+					REGEXP_CHUNKHASH,
+					hashLength(replacer(chunkHash), chunkHashWithLength)
+				)
+				.replace(
+					REGEXP_CONTENTHASH,
+					hashLength(replacer(contentHash), contentHashWithLength)
+				)
+				// query is optional, it's OK if it's in a path
+				// but there's nothing to replace it with
+				.replace(REGEXP_QUERY, replacer(data.query, true))
+				.replace(REGEXP_FILE, replacer(data.filename))
+				.replace(REGEXP_FILEBASE, replacer(data.basename))
+			// TODO
+			// .replace(REGEXP_EXT, replacer(extensions[type]))
 		);
 	}
 
-	return (
-		path
-			.replace(
-				REGEXP_HASH,
-				withHashLength(getReplacer(data.hash), data.hashWithLength)
-			)
-			.replace(
-				REGEXP_CHUNKHASH,
-				withHashLength(getReplacer(chunkHash), chunkHashWithLength)
-			)
-			.replace(
-				REGEXP_CONTENTHASH,
-				withHashLength(getReplacer(contentHash), contentHashWithLength)
-			)
-			.replace(
-				REGEXP_MODULEHASH,
-				withHashLength(getReplacer(moduleHash), moduleHashWithLength)
-			)
-			.replace(REGEXP_ID, getReplacer(chunkId))
-			.replace(REGEXP_MODULEID, getReplacer(prepareId(moduleId)))
-			.replace(REGEXP_NAME, getReplacer(chunkName))
-			.replace(REGEXP_FILE, getReplacer(data.filename))
-			.replace(REGEXP_FILEBASE, getReplacer(data.basename))
-			// query is optional, it's OK if it's in a path but there's nothing to replace it with
-			.replace(REGEXP_QUERY, getReplacer(data.query, true))
-	);
+	// Module Context
+	//
+	// Placeholders
+	//
+	// [id] - module.id (2.png)
+	// [ext] - extname(module.resource || data.filenam) (file.{png, svg, ...})
+	// [hash] - module.hash (6237543873.png)
+	// [name] - file.replace(extname, '') (file)
+	// [file] - basename(module.resource || data.filename) (file.png)
+	//
+	// Legacy Placeholders
+	//
+	// [moduleid] - module.id (2.png)
+	// [modulehash] - module.hash (6237543873.png)
+	if (data.module) {
+		const module = data.module;
+
+		const id =
+			module instanceof Module ? chunkGraph.getModuleId(module) : module.id;
+
+		const hash =
+			module instanceof Module
+				? chunkGraph.getRenderedModuleHash(module)
+				: module.renderedHash || module.hash;
+		// @ts-ignore
+		const hashWithLength = module.hashWithLength || undefined;
+
+		const file = basename(data.filename || "");
+
+		const ext = extname(file).slice(1);
+		const name = file.replace(`.${ext}`, "");
```

Comment:
I think it works better if `ext` includes the `.`. It's similar to `loader-utils` and handles cases where files have no extensions better:

| config | `file.txt` | `some/file.txt` | `some/file` |
|---|---|---|---|
| `[path][name][ext]` | `file.txt` | `some/file.txt` | `some/file` |
| `[path][name].[ext]` | `file.txt` | `some/file.txt` | **`some/file.`** |